pkg/controller/topology.go

```diff
@@ -160,40 +183,78 @@ func aggregateTopologies(
+			return nil, nil
+		}
+
 	} else {
-		// TODO (verult) retry
+		// Delayed binding; scheduler picked selectedNode
+
+		// TODO (#144): use informers
 		selectedNodeInfo, err := kubeClient.StorageV1beta1().CSINodes().Get(selectedNode.Name, metav1.GetOptions{})
 		if err != nil {
-			// We must support provisioning if CSINode is missing, for backward compatibility.
-			klog.Warningf("error getting CSINode for selected node %q: %v; proceeding to provision without topology information", selectedNode.Name, err)
-			return nil, nil
+			// If the Node is before K8s 1.14, then fallback to "topology disabled" behavior
+			nodeVersion, err := version.ParseSemantic(selectedNode.Status.NodeInfo.KubeletVersion)
+			if err != nil {
+				return nil, fmt.Errorf("Failed to parse kubelet version from node %q: %v", selectedNode.Name, err)
+			}
+			if nodeVersion.LessThan(k8sTopologyBetaVersion) {
+				klog.Warningf("Selected node %q version %q is less than %q, falling back to no topology", selectedNode.Name, nodeVersion, k8sTopologyBetaVersion)
+				return nil, nil
+			}
+
+			// Otherwise, CSINode beta feature on the K8s node must also be enabled.
+			// We don't want to fallback and provision in the wrong topology if there's some temporary
+			// error with the API server.
+			return nil, fmt.Errorf("error getting CSINode for selected node %q: %v", selectedNode.Name, err)
 		}
 		topologyKeys = getTopologyKeys(selectedNodeInfo, driverName)
-	}
+		if len(topologyKeys) == 0 {
+			// The scheduler selected a node with no topology information.
+			// This can happen if:
+			//
+			// * the node driver is not deployed on all nodes.
+			// * the node driver is being restarted and has not re-registered yet. This should be
+			//   temporary and a retry should eventually succeed.
+			//
+			// Returning an error in provisioning will cause the scheduler to retry and potentially
+			// (but not guaranteed) pick a different node.
+			return nil, fmt.Errorf("no topology key found on CSINode %s", selectedNode.Name)
+		}
 
-	if len(topologyKeys) == 0 {
-		// Assuming the external provisioner is never running during node driver upgrades.
-		// If selectedNode != nil, the scheduler selected a node with no topology information.
-		// If selectedNode == nil, all nodes in the cluster are missing topology information.
-		// In either case, provisioning needs to be allowed to proceed.
-		return nil, nil
+		// Even though selectedNode is set, we still need to aggregate topology values across
+		// all nodes in order to find secondary topologies.
 	}
 
+	// 2. Find all nodes with the topology keys and extract the topology values
 	selector, err := buildTopologyKeySelector(topologyKeys)
 	if err != nil {
 		return nil, err
 	}
+	// TODO (#144): use informers
 	nodes, err := kubeClient.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: selector})
 	if err != nil {
+		// If the topology feature in the external-provisioner is enabled,
```

Comment:
this comment seems to be repeated a lot. I wonder if there's a better way to capture this sentiment either in the code, error message, or whether it's even necessary in every spot. My one worry is that if we have this comment in each spot its "necessary" right now, what if we add new error conditions and forget to copy the comment, then future maintainers might be like "oh this comment is everywhere except here, maybe it doesn't apply here"...

This is not a big issue imo, but if we can think of a better solution now without too much effort that would be cool :)

Reply:
I think the "If the topology feature is enabled" part is sort of implied, and can be removed